MO-538: Require role to change offender complexity levels

ollietreend opened the pull request:
Jira ticket: MO-538

---

This pull request adds authorisation to the POST endpoint which updates the offender's Complexity of Need level.

The endpoint now requires the client to have role `ROLE_COMPLEXITY_OF_NEED` with scope `write`, as documented in the Swagger definition.

### Summary of changes

- Refactor the auth handling code in `ApplicationController` to make it reusable between `read` and `write` endpoints (use `authorise_read!` or `authorise_write!`)
- Remove redundant check for expired tokens in `HmppsApi::Oauth::Token` – `JWT.decode` already does this for us, and will raise an error if the token has expired
- Add `401 Unauthorized` and `403 Forbidden` response definitions to Swagger

ollietreend committed:
Require write permissions on the POST endpoint

ollietreend committed:
Add auth error responses to Swagger document

starswan commented on app/controllers/application_controller.rb:
```diff
@@ -1,27 +1,47 @@
+                   access_token = auth_header.split.last
+                   HmppsApi::Oauth::Token.new(access_token)
+                 end
+               rescue JWT::DecodeError, NoMethodError
```
why are we rescuing NoMethodError here? - looks scary

starswan replied:
Actually there's a recognised pattern for this - use RescueFrom in the controller, and don't catch it here...

starswan reviewed (commented):
1 issue, but otherwise looks good

ollietreend committed:
Test auth error scenarios on each endpoint

I've pulled some auth-related tests into `shared_examples` blocks in an
attempt to DRY out the tests performed on each endpoint.

I've also added consistent auth error testing on each endpoint. Even
though this auth is performed by the ApplicationController, these tests
prove that it happens on each endpoint.
